crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -2220,6 +2265,10 @@ impl<'a> Checker<'a> {
                     }
                 }
             }
+            // If we're parsing string annotations inside string annotations
+            // (which is the only reason we might enter a second iteration of this loop),
+            // the cache is no longer valid. We must invalidate it to avoid an infinite loop:
+            self.parsed_annotations_cache.clear();
```

Comment:
Can you tell me more about what is inside the cache that is no longer valid? I don't see that this loop modifies the AST in anyway.

Reply:
Sure. Consider the following annotation:

```py
x: "list['str']"
```

The first time we visit the AST< we see `"list['str']"` and identify it as a stringified annotation. We store it in `self.visit.string_type_definitions` to be analyzed later.

After the entire tree has been visited, we look through `self.visit.string_type_definitions` and find `"list['str']"`. We parse it, and it becomes `list['str']`. After parsing it, we call `self.visit_expr()` on the `list['str']` node, and that `visit_expr` call is going to find `'str'` inside that node and identify it as a string type definition, appending it to `self.visit.string_type_definitions`, ensuring that there will be one more iteration of this loop. Unfortunately, the `TextRange` of `'str'` here will be _relative to the parsed `list['str']` node_ rather than _relative to the original module_, meaning the cache (which uses `TextSize` as the key) becomes invalid on the second iteration of this loop.

Does that make sense? Can you see a better way round this?

Reply:
(The effect of this is that without clearing the cache on each iteration of this loop, we can have an infinite loop on some snippets. I observed this when running `cargo test -p ruff_linter` on an early version of this PR, and it took a while to debug 😆)

Reply:
Makes sense. It might make sense to paste this excellent explanation into the comment :)

No, I don't have another workaround other than using ptr-comparison and storing expression references similar to https://github.com/astral-sh/ruff/blob/d6851076388624d4c76c68d0cf8e72fd29e4a6e6/crates/ruff_python_formatter/src/comments/node_key.rs#L9-L12

Reply:
> Makes sense. It might make sense to paste this excellent explanation into the comment :)

I'll do so!